core/src/main/java/org/elasticsearch/script/Script.java

```diff
@@ -32,268 +31,545 @@
+                            } else {
+                                throw new ParsingException(parser.getTokenLocation(),
+                                    "unexpected value [" + (token.isValue() ? parser.text() : token) + "], expected [<option value>]");
+                            }
+                        } else {
+                            throw new ParsingException(parser.getTokenLocation(),
+                                "unexpected value [" + (token.isValue() ? parser.text() : token) + "], expected [<option name>]");
+                        }
+                    }
+                } else {
+                    throw new ParsingException(parser.getTokenLocation(),
+                        "unexpected value [" + (token.isValue() ? parser.text() : token) + "], expected [<options>]");
+                }
+            } else if (matcher.match(name, ScriptField.PARAMS)) {
+                if (token == Token.START_OBJECT) {
+                    params = parser.map();
+                } else {
+                    throw new ParsingException(parser.getTokenLocation(),
+                        "unexpected value [" + (token.isValue() ? parser.text() : token) + "], expected [<params>]");
+                }
+            } else {
+                throw new ParsingException(parser.getTokenLocation(),
+                    "unexpected field [" + (name == null ? parser.currentToken() : name) + "], " + "expected [" +
+                        ScriptType.INLINE.getParseField().getPreferredName() + ", " +
+                        ScriptType.STORED.getParseField().getPreferredName() + ", " +
+                        ScriptType.FILE.getParseField().getPreferredName() + ", " +
+                        ScriptField.LANG.getPreferredName() + ", " +
+                        ScriptField.PARAMS.getPreferredName() +
+                        "]");
+            }
+        }
+
+        if (type == null) {
+            throw new IllegalArgumentException("missing required parameter for parsing script, expected valid script type [" +
+                ScriptType.INLINE.getName() + "] or [" + ScriptType.STORED.getName() + "] or [" + ScriptType.FILE.getName() + "]");
+        }
+
+        if (idOrCode == null) {
+            throw new IllegalArgumentException("missing required parameter for parsing script, expected [<id>] or [<code>]");
+        }
+
+        if (params == null) {
+            params = new HashMap<>();
+        }
+
+        return new Script(type, lang, idOrCode, options, params);
+    }
+
     @SuppressWarnings("unchecked")
-    public Script(String script, ScriptType type, String lang, @Nullable Map<String, ?> params,
-                  @Nullable XContentType  contentType) {
-        if (contentType != null && type != ScriptType.INLINE) {
-            throw new IllegalArgumentException("The parameter contentType only makes sense for inline scripts");
+    public static Script readFrom(StreamInput in) throws IOException {
```

Comment:
would it make sense to remove this readFrom method and add `public Script(StreamInput input)` this is the pattern we use with all writeables?